ruby/ext/google/protobuf_c/storage.c

```diff
@@ -178,9 +178,42 @@ void native_slot_set_value_and_case(const char* name,
       if (CLASS_OF(value) == CLASS_OF(Qnil)) {
         value = Qnil;
       } else if (CLASS_OF(value) != type_class) {
-        rb_raise(cTypeError,
-                 "Invalid type %s to assign to submessage field '%s'.",
-                rb_class2name(CLASS_OF(value)), name);
+        // check for possible implicit conversions
+        VALUE converted_value = NULL;
+        char* field_type_name = rb_class2name(type_class);
+        char* value_type_name = rb_class2name(CLASS_OF(value));
+
+        // Time -> Google::Protobuf::Timestamp
+        if (strcmp(field_type_name, "Google::Protobuf::Timestamp") == 0 &&
+            strcmp(value_type_name, "Time") == 0) {
+          VALUE hash = rb_hash_new();
+          rb_hash_aset(hash, rb_str_new2("seconds"), rb_funcall(value, rb_intern("to_i"), 0));
+          rb_hash_aset(hash, rb_str_new2("nanos"), rb_funcall(value, rb_intern("nsec"), 0));
+          VALUE args[1] = { hash };
+          converted_value = rb_class_new_instance(1, args, type_class);
```

Comment:
Get out  of this branch here?

Reply:
Sorry, I'm not sure what you mean by that. Could you explain?

Reply:
I mean, after you have finished processing Timestamp, you don't need to check whether it's Duration.

Reply:
Ah, you mean move it to an `else if`. Done.